share/eds/p2p/client.go

```diff
@@ -0,0 +1,161 @@
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+
+	"github.com/celestiaorg/go-libp2p-messenger/serde"
+
+	"github.com/celestiaorg/celestia-node/share"
+	"github.com/celestiaorg/celestia-node/share/eds"
+	p2p_pb "github.com/celestiaorg/celestia-node/share/eds/p2p/pb"
+
+	"github.com/celestiaorg/rsmt2d"
+)
+
+var log = logging.Logger("shrex/eds")
+
+var errNoMorePeers = errors.New("all peers returned invalid responses")
+
+// TODO(@distractedm1nd): make version suffix configurable
+var protocolID = protocol.ID("/shrex/eds/0.0.1")
+
+// Client is responsible for requesting EDSs for blocksync over the ShrEx/EDS protocol.
+// This client is run by light nodes and full nodes. For more information, see ADR #13
+type Client struct {
+	protocolID protocol.ID
+	host       host.Host
+}
+
+// NewClient creates a new ShrEx/EDS client.
+func NewClient(host host.Host) *Client {
+	return &Client{
+		host:       host,
+		protocolID: protocolID,
+	}
+}
+
+// RequestEDS requests the full EDS from one of the given peers.
+//
+// The peers are requested in a round-robin manner with retries until one of them gives a valid
+// response. Blocks forever until the context is canceled or a valid response is given.
+func (c *Client) RequestEDS(
+	ctx context.Context,
+	dataHash share.DataHash,
+	peers peer.IDSlice,
+) (*rsmt2d.ExtendedDataSquare, error) {
+	req := &p2p_pb.EDSRequest{Hash: dataHash}
+
+	// requests are retried for every peer until a valid response is received
+	excludedPeers := make(map[peer.ID]struct{})
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+		}
+
+		// if no peers are left, return
+		if len(peers) == len(excludedPeers) {
+			return nil, errNoMorePeers
+		}
+
+		for _, to := range peers {
+			select {
+			case <-ctx.Done():
+				return nil, ctx.Err()
+			default:
+			}
+
+			// skip over excluded peers
+			if _, ok := excludedPeers[to]; ok {
+				continue
+			}
+
+			eds, err := c.doRequest(ctx, req, dataHash, to)
+			if err != nil {
+				// peer has misbehaved, exclude them from round-robin
+				excludedPeers[to] = struct{}{}
+				log.Errorw("client: eds request to peer failed", "peer", to, "hash", dataHash.String())
+			}
+
+			// eds is nil when the request was valid but couldn't be served
+			if eds != nil {
+				return eds, nil
+			}
+		}
+	}
+}
+
+func (c *Client) doRequest(
+	ctx context.Context,
+	req *p2p_pb.EDSRequest,
+	dataHash share.DataHash,
+	to peer.ID,
+) (*rsmt2d.ExtendedDataSquare, error) {
+	log.Debugf("client: requesting eds %s from peer %s", dataHash.String(), to)
+	stream, err := c.host.NewStream(ctx, to, c.protocolID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open stream: %w", err)
+	}
+
+	// request status from peer to provide the requested EDS
+	err = stream.SetWriteDeadline(time.Now().Add(writeDeadline))
```

Comment:
To follow what we've discussed in HeaderEx, the deadlines should be set from the context. Otherwise context cancellation does not take any effect and blocks forever

Reply:
done, but server side still uses durations yeah? Since there is no context on handleStream

Reply:
Yes, server uses durations.